File: src/MixNetFx/FacessoData/SQLServer/SPAccess_WageGroups.vb
Imports Activedev
Imports Facesso
Imports System.Data
Imports System.Data.SqlClient
Imports System.Data.Common

Partial Public NotInheritable Class SPAccess

    Public Function WageGroups_DoesTokenExist(ByVal IDSubsidiary As Integer, _
                                              ByVal WageGroupToken As String, _
                                              ByVal ExcludeIDWageGroup As ADDBNullable(Of Integer)) _
                                              As Boolean

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("WageGroups_DoesTokenExist", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = IDSubsidiary
                .Add("@WageGroupToken", SqlDbType.NVarChar, 20).Value = WageGroupToken
                .Add("@ExcludeIDWageGroup", SqlDbType.Int).Value = ExcludeIDWageGroup.Value
                .Add("@DoesExist", SqlDbType.Bit)
                .Item("@DoesExist").Direction = ParameterDirection.Output
            End With
            locCmd.ExecuteReader()
            Return CBool(locCmd.Parameters("@DoesExist").Value)
        End Using
    End Function

    Public Function WageGroups_Add(ByVal wgi As WageGroupInfo, ByVal CreatedByIDUser As Integer) As Integer

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("WageGroups_Add", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = wgi.IDSubsidiary
                .Add("@IDCurrency", SqlDbType.Int).Value = wgi.IDCurrency
                .Add("@IsTemplate", SqlDbType.Bit).Value = wgi.IsTemplate
                .Add("@WageGroupToken", SqlDbType.NVarChar, 20).Value = wgi.WageGroupToken
                .Add("@Comment", SqlDbType.NVarChar, 4000).Value = wgi.Comment.Value
                .Add("@HourlyRate", SqlDbType.Money).Value = wgi.HourlyRate
                .Add("@WasCurrentTo", SqlDbType.DateTime).Value = FacessoGeneric.OpenCurrentToDate
                .Add("@CreatedByIDUser", SqlDbType.Int).Value = CreatedByIDUser
                .Add("@IDWageGroupNew", SqlDbType.Int)
                .Item("@IDWageGroupNew").Direction = ParameterDirection.Output
            End With
            locCmd.ExecuteReader()
            Return CInt(locCmd.Parameters("@IDWageGroupNew").Value)
        End Using
    End Function

    Public Function WageGroups_Edit(ByVal wgi As WageGroupInfo, ByVal LastEditedByIDUser As Integer) As Integer

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("WageGroups_Edit", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = wgi.IDSubsidiary
                .Add("@IDCurrency", SqlDbType.Int).Value = wgi.IDCurrency
                .Add("@IDWageGroup", SqlDbType.Int).Value = wgi.IDWageGroup
                .Add("@IsTemplate", SqlDbType.Bit).Value = wgi.IsTemplate
                .Add("@WageGroupToken", SqlDbType.NVarChar, 20).Value = wgi.WageGroupToken
                .Add("@Comment", SqlDbType.NVarChar, 4000).Value = wgi.Comment.Value
                .Add("@HourlyRate", SqlDbType.Money).Value = wgi.HourlyRate
                .Add("@LastEditedByIDUser", SqlDbType.Int).Value = LastEditedByIDUser
                .Add("@ConsiderHistoryMaintenance", SqlDbType.Bit).Value = ConsiderHistoryMaintenance
                .Add("@IDWageGroupNew", SqlDbType.Int)
                .Item("@IDWageGroupNew").Direction = ParameterDirection.Output
            End With
            locCmd.ExecuteReader()
            Return CInt(locCmd.Parameters("@IDWageGroupNew").Value)
        End Using
    End Function

    Public ReadOnly Property WageGroupInfoCollection() As WageGroupInfoCollection
        Get
            Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
            If locConnection Is Nothing Then
                Return Nothing
            End If
            Using locConnection
                Dim locCommand As New SqlCommand( _
                "SELECT [WageGroups].*,[Currencies].[CurrencyToken] From [WageGroups] " & _
                "[WageGroups] INNER JOIN [Currencies] ON " & _
                "[WageGroups].[IDCurrency] = [Currencies].[IDCurrency] WHERE " & _
                "[WageGroups].[IDSubsidiary]=" & FacessoGeneric.LoginInfo.IDSubsidiary & _
                " AND [WageGroups].[IsCurrent]='true'", locConnection)
                Dim locDR As SqlDataReader = locCommand.ExecuteReader()
                If locDR.HasRows Then
                    Dim locWic As New WageGroupInfoCollection
                    Do While locDR.Read
                        Dim locWageGroupInfo As New WageGroupInfo(locDR, True)
                        locWic.Add(locWageGroupInfo)
                    Loop
                    Return locWic
                Else
                    Return Nothing
                End If
            End Using
        End Get
    End Property

    Public Function GetWageGroup(ByVal IDSubsidiary As Integer, ByVal IDWageGroup As Integer) As WageGroupInfo
        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCommand As New SqlCommand( _
            "SELECT [WageGroups].*,[WageGroups].[CurrencyToken] From [WageGroups] " & _
            "[WageGroups] INNER JOIN [Currencies] ON " & _
            "[WageGroups].[IDCurrency] = [Currencies].[IDCurrency] WHERE " & _
            "[WageGroups].[IDSubsidiary]=" & FacessoGeneric.LoginInfo.IDSubsidiary & _
            " AND [WageGroups].[IDWageGroups]=" & IDWageGroup, locConnection)
            Dim locDR As SqlDataReader = locCommand.ExecuteReader()
            If locDR.HasRows Then
                locDR.Read()
                Return New WageGroupInfo(locDR, True)
            Else
                Return Nothing
            End If
        End Using
    End Function
End Class

File: src/MixNetFx/Facesso.Data/SQLServer/SPAccess_WageGroups.cs
using System;
using System.Data;
using System.Data.SqlClient;
using ActiveDev;
using Facesso;

namespace Facesso.Data
{
    public sealed partial class SPAccess
    {
        public bool WageGroups_DoesTokenExist(int idSubsidiary, string wageGroupToken, ADDBNullable<int> excludeIDWageGroup)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return false;
            using (locConnection)
            {
                var locCmd = new SqlCommand("WageGroups_DoesTokenExist", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = idSubsidiary;
                locCmd.Parameters.Add("@WageGroupToken", SqlDbType.NVarChar, 20).Value = wageGroupToken;
                locCmd.Parameters.Add("@ExcludeIDWageGroup", SqlDbType.Int).Value = excludeIDWageGroup.Value;
                locCmd.Parameters.Add("@DoesExist", SqlDbType.Bit);
                locCmd.Parameters["@DoesExist"].Direction = ParameterDirection.Output;
                locCmd.ExecuteReader();
                return (bool)locCmd.Parameters["@DoesExist"].Value;
            }
        }

        public int WageGroups_Add(WageGroupInfo wgi, int createdByIDUser)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return 0;
            using (locConnection)
            {
                var locCmd = new SqlCommand("WageGroups_Add", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = wgi.IDSubsidiary;
                locCmd.Parameters.Add("@IDCurrency", SqlDbType.Int).Value = wgi.IDCurrency;
                locCmd.Parameters.Add("@IsTemplate", SqlDbType.Bit).Value = wgi.IsTemplate;
                locCmd.Parameters.Add("@WageGroupToken", SqlDbType.NVarChar, 20).Value = wgi.WageGroupToken;
                locCmd.Parameters.Add("@Comment", SqlDbType.NVarChar, 4000).Value = wgi.Comment.Value;
                locCmd.Parameters.Add("@HourlyRate", SqlDbType.Money).Value = wgi.HourlyRate;
                locCmd.Parameters.Add("@WasCurrentTo", SqlDbType.DateTime).Value = FacessoGeneric.OpenCurrentToDate;
                locCmd.Parameters.Add("@CreatedByIDUser", SqlDbType.Int).Value = createdByIDUser;
                locCmd.Parameters.Add("@IDWageGroupNew", SqlDbType.Int);
                locCmd.Parameters["@IDWageGroupNew"].Direction = ParameterDirection.Output;
                locCmd.ExecuteReader();
                return (int)locCmd.Parameters["@IDWageGroupNew"].Value;
            }
        }

        public int WageGroups_Edit(WageGroupInfo wgi, int lastEditedByIDUser)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return 0;
            using (locConnection)
            {
                var locCmd = new SqlCommand("WageGroups_Edit", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = wgi.IDSubsidiary;
                locCmd.Parameters.Add("@IDCurrency", SqlDbType.Int).Value = wgi.IDCurrency;
                locCmd.Parameters.Add("@IDWageGroup", SqlDbType.Int).Value = wgi.IDWageGroup;
                locCmd.Parameters.Add("@IsTemplate", SqlDbType.Bit).Value = wgi.IsTemplate;
                locCmd.Parameters.Add("@WageGroupToken", SqlDbType.NVarChar, 20).Value = wgi.WageGroupToken;
                locCmd.Parameters.Add("@Comment", SqlDbType.NVarChar, 4000).Value = wgi.Comment.Value;
                locCmd.Parameters.Add("@HourlyRate", SqlDbType.Money).Value = wgi.HourlyRate;
                locCmd.Parameters.Add("@LastEditedByIDUser", SqlDbType.Int).Value = lastEditedByIDUser;
                locCmd.Parameters.Add("@ConsiderHistoryMaintenance", SqlDbType.Bit).Value = FacessoGeneric.ConsiderHistoryMaintenance;
                locCmd.Parameters.Add("@IDWageGroupNew", SqlDbType.Int);
                locCmd.Parameters["@IDWageGroupNew"].Direction = ParameterDirection.Output;
                locCmd.ExecuteReader();
                return (int)locCmd.Parameters["@IDWageGroupNew"].Value;
            }
        }

        public WageGroupInfoCollection WageGroupInfoCollection
        {
            get
            {
                SqlConnection locConnection = GetOpenedConnectionSafely();
                if (locConnection == null) return null;
                using (locConnection)
                {
                    var locCommand = new SqlCommand(
                        "SELECT [WageGroups].*,[Currencies].[CurrencyToken] From [WageGroups] " +
                        "[WageGroups] INNER JOIN [Currencies] ON " +
                        "[WageGroups].[IDCurrency] = [Currencies].[IDCurrency] WHERE " +
                        "[WageGroups].[IDSubsidiary]=" + FacessoGeneric.LoginInfo.IDSubsidiary +
                        " AND [WageGroups].[IsCurrent]='true'", locConnection);
                    SqlDataReader locDR = locCommand.ExecuteReader();
                    if (locDR.HasRows)
                    {
                        var locWic = new WageGroupInfoCollection();
                        while (locDR.Read())
                        {
                            var locWageGroupInfo = new WageGroupInfo(locDR, true);
                            locWic.Add(locWageGroupInfo);
                        }
                        return locWic;
                    }
                    return null;
                }
            }
        }

        public WageGroupInfo GetWageGroup(int idSubsidiary, int idWageGroup)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return null;
            using (locConnection)
            {
                var locCommand = new SqlCommand(
                    "SELECT [WageGroups].*,[WageGroups].[CurrencyToken] From [WageGroups] " +
                    "[WageGroups] INNER JOIN [Currencies] ON " +
                    "[WageGroups].[IDCurrency] = [Currencies].[IDCurrency] WHERE " +
                    "[WageGroups].[IDSubsidiary]=" + FacessoGeneric.LoginInfo.IDSubsidiary +
                    " AND [WageGroups].[IDWageGroups]=" + idWageGroup, locConnection);
                SqlDataReader locDR = locCommand.ExecuteReader();
                if (locDR.HasRows)
                {
                    locDR.Read();
                    return new WageGroupInfo(locDR, true);
                }
                return null;
            }
        }
    }
}
